### PR TITLE
Document events

### PR DESCRIPTION
**Pull request by ghost**
As a follow-up to #10322 and #10321. I have done some event documentation, just to see as this would look like.

Should we go that way and add event documentation over the time, when we are touching an event handler? 
(Please no nit-picking here, just a general direction)

grafik

grafik

grafik

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/10328)

### CODE REVIEW

**Comment by poire-z** [comment]
I personally have no real interest nor expectations regarding our documentation - so I won't bother you about it :)

... but still :)

For such as this one, I dunno if it's worth ending in the module documentation, it's mostly just code comment explaining what it does and why:
```lua
---
-- We do not want AutoTurn to turn pages during the suspend process.
--
-- Unschedule it and restart after resume.
-- @event Suspend
function AutoTurn:onSuspend()
```
but well, it might be just me that discovered docs when @Frenzie requested me to add docs to widgets (or Trapper, one I added a lot to) - and for widgets that should be reused, I can understand some comments ending up as doc may help. For an upper level widget like BookMap or a plugin, which is more a standalone application that something that can be reused, it makes less sense to get all its comments end up as docs.

Something that I would find more helpful, about events, would be some cross refs of all event sents (handleEvent/broadcastEvent/...), the locations that send them - and the list of all handlers and their locations (without the need to add `@doctags` to any of them, just parsing `:onXYZ`), so it's easier to know what reacts to what and the code to look at to know more and not forget when you change something. Although I'm fine using grep when I need that.

**Comment by NiLuJe** [comment]
> Should we go that way and add event documentation over the time, when we are touching an event handler?

That sounds good to me; Events are good things to know about and track for potential contributors (or just groking how something works, really ;p).

And that somewhat tracks with what @poire-z said, what's (mostly) of practical interest is *where* they're sent and where they're handled (i.e., another entry in the long series of `ripgrep` is my best friend). (I mean, a vague overview of what they're for is also helpful, but, ideally, we take care of that by using a descriptive name first and foremost ;)).

I vaguely remember that we can reference modules without too much manual intervention, but I don't LuaDoc enough to say if we could automate something like that or not.

**Comment by yparitcher** [comment]
The idea is good as events are our real IPC method.

It is not useful to document what each module does with a generic event (we don't care how each module handles suspend, wifi etc).

What i think is important as stated above is a reference of each place a event is sent and handled, a general summary of what the event is for (its name may be enough, `onSuspend` is descriptive, some of the wifi ones might want some more differentiation) 

For specific events that are more specific (eg `onToggleMightMode`) we could document the handler, and auto list all callers

**Comment by Frenzie** [comment]
I'm not overly fond of it. I find it more useful to ensure the greppability of the code (i.e., write out all the onX, onY, onZ and don't do some kind of autogenerated on{X,Y,Z}).

Incidentally @poire-z in editors with LSP support you can run e.g. <https://github.com/LuaLS/lua-language-server> and get extra fancy IDE autocomplete stuff from our documentation. As I've said before I use Kate and Geany, but it's easier to set up the LSP stuff in VS Code for this quick screenshot.
Screenshot_20230415_000244

**Comment by Frenzie** [comment]
I'm not entirely sure what I'm looking at (or perhaps rather, the result of what), but yeah, that looks like it could be interesting.

**Comment by Frenzie** [comment]
> make some cross references (I would like something like https://github.com/see ???, let's see what ldoc is capable off)

gettext is probably the most full-featured one I wrote

> don't call ldoc twice in the Makefile

That might be unavoidable. The resulting file ultimately shouldn't be committed btw.

**Comment by poire-z** [comment]
> A href link to the source code (I am looking for a method to open that file on a certain line number).

```
https://github.com/koreader/koreader/blob/master/frontend/apps/reader/modules/readerfooter.lua#L2507
```
Not good enough ? :)

**Comment by Frenzie** [comment]
It's not, <https://github.com/koreader/koreader/blob/99265ba1b7061ae0521be8e777db8bdfbf59c2dc/frontend/apps/reader/modules/readerfooter.lua#L2507> would be better for the potential 20-minute delay. ;-)

Btw, could you explain what the kodoc.css thing is for? It looks like the exact same thing as the default basically?

**Comment by Frenzie** [comment]
> The kodoc.css/ldoc.css is basically the file shipped with ldoc, but it changes/increases the width if the window is wider than the 700px.

There's no method to ship a five-line additional CSS file? Having a whole theme that can get out of sync with upstream seems a bit suboptimal. :-/

**Comment by NiLuJe** [comment]
That looks good to me ;).

**Comment by poire-z** [comment]
It looks like a nice reference sheet - and should be handy the 2 or 3 times (only) each year I have to tweak/refactor some event/handler bits (last time was with the cre partial rerenderings).

If it would not need these `@event` and `@eventHandler` annotations, I would trust it even more - but the fact that it needs it, and that I (and others) will forget to add them (or failt to notice they are missing when reviewing) - I might end up not fully trusting this sheet, and resort to grep to be really sure I haven't missed any.
So, nice indeed to get an overall look of what is at play - but not fully trustable or strictly necessary :/

But if it doesn't need these annotations, and even if it catches false positives (as long as it doesn't miss any), yes, good to have.

(Also, dunno if you noticed these, but there are also handlers aliased to other handlers:
```
$ rgrep '\.on.*=' frontend/apps/reader/modules/
frontend/apps/reader/modules/readerthumbnail.lua:ReaderThumbnail.onBookmarkEdited = ReaderThumbnail.resetCachedPagesForBookmarks
frontend/apps/reader/modules/readerfont.lua:ReaderFont.onPhysicalKeyboardConnected = ReaderFont.registerKeyEvents
frontend/apps/reader/modules/readerpaging.lua:ReaderPaging.onPhysicalKeyboardConnected = ReaderPaging.registerKeyEvents
frontend/apps/reader/modules/readerrolling.lua:ReaderRolling.onPhysicalKeyboardConnected = ReaderRolling.registerKeyEvents
frontend/apps/reader/modules/readerback.lua:ReaderBack.onPhysicalKeyboardConnected = ReaderBack.registerKeyEvents
frontend/apps/reader/modules/readerback.lua:ReaderBack.onReaderReady = ReaderBack._onViewPossiblyUpdated
```

**Comment by Frenzie** [comment]
@zwim 
It looks good. Perhaps I might use it more in time, though as I said from the onset I'm more than satisfied with (rip)grep and similar. My main concern is the same as @poire-z's, namely that it looks quite manual in this PR, much more so than I had expected.

@poire-z 
> If it would not need these `@event` and `@eventHandler` annotations, I would trust it even more - but the fact that it needs it, and that I (and others) will forget to add them (or failt to notice they are missing when reviewing) - I might end up not fully trusting this sheet, and resort to grep to be really sure I haven't missed any.

On that note, if luacheck itself doesn't provide a way to do this without annotations (and it probably doesn't but I didn't look very hard), perhaps it can simply be done in two steps.

1. "grep" to auto-annotate
2. What's there now.

The auto-annotation could be done as a pre-commit hook if there's value to having it in the code, or it could just be there temporarily for doc generation only. Though then I suppose you'd need something like git stash, auto-annotate, git reset, git stash pop so you can still do it locally.

**Comment by Frenzie** [comment]
> A simple solution would copy the whole frontend-tree to a tmp directory (6MB) and annotate everything (=add comments) there. We could use this (annotated)frontend for make doc.

That reads to me like a worse version of what I said above ("git stash, auto-annotate, git reset, git stash pop")? ^_^ Btw, `base` is also included in the docs.

**Comment by Frenzie** [comment]
There are flags like -a, -i, and probably others.

**Comment by Frenzie** [comment]
> A minor back draw exists: If the module documentation has to be added, the link in the Events and hander-page is wrong by three lines.

Using a multiline comment there is the better stylistic choice in normal circumstances, but it would work all the same with just `---`.

File: frontend/ui/widget/bboxwidget.lua [hunk 1]
```diff
@@ -323,7 +323,7 @@ function BBoxWidget:onSwipeAdjust(arg, ges)
 end
 
 function BBoxWidget:onHoldAdjust(arg, ges)
-    --- @fixme this is a dirty hack to disable hold gesture in page cropping
```
Comment (Frenzie) [inline frontend/ui/widget/bboxwidget.lua:326]: What's the problem with this precisely?

**Comment by NiLuJe** [comment]
Yeah (also, I actually rely on onCloseWidget a lot, so having an overview of it *is* helpful; to me, at least ;p).

**Comment by yparitcher** [comment]
It looks good.
I am on mobile so i could not see it styled, but i assume it is in a table like the rest of the docs. If so can the cell borders in between one event+handler group and the next be thicker? This would help looking through.
Thanks

**Comment by NiLuJe** [comment]
> This is how it looks like now: 

Stupid question: why are some of the horizontal separators thinner?

**Comment by Frenzie** [comment]
It only creates a page without otherwise affecting the docs, is that correct?

File: plugins/autodim.koplugin/main.lua [hunk 2]
```diff
@@ -222,9 +234,12 @@ function AutoDim:clearEventHandlers()
     self.onSuspend = nil
 end
 
+--- Reschedule AutoDim task.
+--
+-- This might be happening through AutoWarmth during a ramp down.
+-- Set original intensity, but don't turn fl on actually.
+-- @eventHandler onFrontlightTurnedOff
 function AutoDim:onFrontlightTurnedOff()
```
Comment (poire-z) [inline plugins/autodim.koplugin/main.lua:242]: I don't remember what we settled on :/
Will we have to add such `@eventhandler onXYZ` above each such method, like you did here?
But you didn't in the chunk below AutoStandby:onAllowStandby() - and it seems if it was really need this PR would have much more of them :) So, was it from some early idea and we don't need them?

**Comment by NiLuJe** [comment]
Still on my TODO list (low prio, but still in it ;)).

**Pull request closed without merging**